core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -106,6 +107,20 @@ final case class NonEmptyList[A](head: A, tail: List[A]) {
     toList.iterator.map(A.show).mkString("NonEmptyList(", ", ", ")")
 
   override def toString: String = s"NonEmpty$toList"
+
+  /**
+   * Remove duplicates. Duplicates are checked using `Order[_]` instance.
+   */
+  def distinct(implicit O: Order[A]): NonEmptyList[A] = {
```

Comment:
> This is perhaps best served with different method names so readers can be more clear which complexity they get

I strongly agree with this statement. Especially since IntelliJ has a nasty habit of telling people that imports are unused when they are only used to bring in implicit instances. It would be really unfortunate if deleting an import brought the `Hash` or `Order` instance out of scope and some code went from `O(n)` to `O(n^2)`.

I don't feel a strong need to add an `O(n^2)` version that requires `Eq[A]` instead of `Order[A]`, because `O(n^2)` operations are often impractical, and I suspect that most of the cases in which you want to do something like this you probably can have an `Order[A]` available. I could definitely see some value in a hash-based approach, but that's a higher barrier given what's currently in Cats.

Personally I'm pretty happy to go forward with this approach and leave open the possibility of another approach in the future.